src/react/apiBase.ts

```diff
@@ -81,67 +66,39 @@ export function useStoreMapBase<State, Result, Keys extends ReadonlyArray<any>>(
   if (!is.store(store)) throwError('useStoreMap expects a store')
   if (!Array.isArray(keys)) throwError('useStoreMap expects an array as keys')
   if (typeof fn !== 'function') throwError('useStoreMap expects a function')
-  const result = React.useRef<{
-    fn: (state: State, keys: Keys) => Result
-    upd: (update: Result, current: Result) => boolean
-    val: Result
-    init: boolean
-    store: Store<State>
-    active: boolean
-  }>({} as any)
-  const refState = result.current
-  refState.fn = fn
-  refState.upd = updateFilter
-  refState.init = refState.store === store
-  refState.store = store
-  refState.active = true
-  useIsomorphicLayoutEffect(
-    () => () => {
-      if (result.current) {
-        result.current.active = false
+
+  const subscribe = (cb: () => void) => createWatch(store, cb, scope)
+  const read = () => stateReader(store, scope)
+
+  const stateRef = React.useRef<State>()
+  const valueRef = React.useRef<Result>()
+  const keysRef = React.useRef(keys)
+
+  const value = useSyncExternalStoreWithSelector(
+    subscribe,
+    read,
+    read,
+    state => {
+      if (stateRef.current !== state || !keysEqual(keysRef.current, keys)) {
+        const result = fn(state, keys)
+
+        stateRef.current = state
+        keysRef.current = keys
+
+        // skip update, if undefined
+        // just like original store or previous implementation
+        if (result !== undefined) {
+          valueRef.current = result
+        }
       }
+
+      return valueRef.current as Result
     },
-    [],
+    // `updateFilter` always had (next, prev) arguments, but `isEqual` of this hook has (prev, next) order of arguments
+    (a, b) => !updateFilter(b, a),
```

Comment:
We can rename `a` and `b` to `prev` and `next`, subsequently, the comment will be extraneous 🤗